### PR TITLE
feat (api): workflow hook model

### PR DESCRIPTION
**Pull request by fsamin**

### CODE REVIEW

File: engine/api/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -52,5 +54,10 @@ func InitiliazeDB(defaultValues DefaultValues, DBFunc func() *gorp.DbMap) error
 		return sdk.WrapError(err, "InitiliazeDB> Cannot setup builtin environments")
 	}
 
+	if err := workflow.CreateBuiltinWorkflowHookModels(dbGorp); err != nil {
+		log.Error("InitiliazeDB> CreateBuiltinWorkflowHookModels failed: %v", err)
```
Comment (yesnault) [inline engine/api/bootstrap/bootstrap.go]: log really needed here ?

File: engine/api/workflow/dao_hook_model.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+		Type:       sdk.WorkflowHookModelBuiltin,
+		Identifier: sdk.WorkflowHookModelBuiltin,
+		Name:       "WebHook",
+		Icon:       "",
+	}
+
+	GitPollerModel = &sdk.WorkflowHookModel{
+		Author:     "CDS",
+		Type:       sdk.WorkflowHookModelBuiltin,
+		Identifier: sdk.WorkflowHookModelBuiltin,
```
Comment (sguiheux) [inline engine/api/workflow/dao_hook_model.go]: same identifier?

File: engine/api/workflow/dao_hook_model.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+		Name:       "Git Repository Poller",
+		Icon:       "",
+	}
+
+	SchedulerModel = &sdk.WorkflowHookModel{
+		Author:     "CDS",
+		Type:       sdk.WorkflowHookModelBuiltin,
+		Identifier: sdk.WorkflowHookModelBuiltin,
```
Comment (sguiheux) [inline engine/api/workflow/dao_hook_model.go]: same identifier?

File: engine/api/workflow_hook.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	defer tx.Rollback()
+
+	if err := workflow.InsertHookModel(tx, m); err != nil {
+		return sdk.WrapError(err, "postWorkflowHookModelHandler")
+	}
+
+	if err := tx.Commit(); err != nil {
+		return sdk.WrapError(errtx, "postWorkflowHookModelHandler> Unable to commit transaction")
```
Comment (yesnault) [inline engine/api/workflow_hook.go]: errtx -> err

File: engine/api/workflow/dao_hook_model.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		return err
+	}
+	return nil
+}
+
+//CreateBuiltinWorkflowHookModels insert all builtin hook models in database
+func CreateBuiltinWorkflowHookModels(db *gorp.DbMap) error {
+	tx, err := db.Begin()
+	if err != nil {
+		return err
```
Comment (sguiheux) [inline engine/api/workflow/dao_hook_model.go]: wrap error

File: engine/api/workflow_hook.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+	return WriteJSON(w, r, m, http.StatusOK)
+}
+
+func getWorkflowHookModelHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *businesscontext.Ctx) error {
+	vars := mux.Vars(r)
+	name := vars["model"]
+	m, err := workflow.LoadHookModelByName(db, name)
+	if err != nil {
+		return sdk.WrapError(err, "getWorkflowHookModelsHandler")
```
Comment (yesnault) [inline engine/api/workflow_hook.go]: getWorkflowHookModelsHandler -> getWorkflowHookModelHandler

File: engine/api/workflow/dao_hook_model.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	defer tx.Rollback()
+
+	if _, err := tx.Exec("LOCK TABLE workflow_hook_model IN ACCESS EXCLUSIVE MODE"); err != nil {
+		return err
```
Comment (sguiheux) [inline engine/api/workflow/dao_hook_model.go]: wrap error

File: engine/api/workflow/dao_hook_model.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	for _, h := range builtinModels {
+		ok, err := checkBuiltinWorkflowHookModelExist(tx, h)
+		if err != nil {
+			return err
```
Comment (sguiheux) [inline engine/api/workflow/dao_hook_model.go]: wrap error

File: engine/api/workflow_hook.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+func putWorkflowHookModelHandler(w http.ResponseWriter, r *http.Request, db *gorp.DbMap, c *businesscontext.Ctx) error {
+	m := &sdk.WorkflowHookModel{}
+	if err := UnmarshalBody(r, m); err != nil {
+		return err
+	}
+
+	tx, errtx := db.Begin()
+	if errtx != nil {
+		return sdk.WrapError(errtx, "postWorkflowHookModelHandler> Unable to start transaction")
```
Comment (yesnault) [inline engine/api/workflow_hook.go]: postWorkflowHookModelHandler -> putWorkflowHookModelHandler

File: engine/api/workflow_hook.go [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	defer tx.Rollback()
+
+	if err := workflow.UpdateHookModel(tx, m); err != nil {
+		return sdk.WrapError(err, "postWorkflowHookModelHandler")
```
Comment (yesnault) [inline engine/api/workflow_hook.go]: put

File: engine/api/workflow_hook.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	if err := tx.Commit(); err != nil {
+		return sdk.WrapError(errtx, "postWorkflowHookModelHandler> Unable to commit transaction")
```
Comment (yesnault) [inline engine/api/workflow_hook.go]: put

File: engine/api/workflow/dao_node.go [hunk 12]
```diff
@@ -224,7 +224,7 @@ func loadNodeContext(db gorp.SqlExecutor, wn *sdk.WorkflowNode, u *sdk.User) (*s
 		}
 	}
 
-	//Load the application in the context
+	//Load the application in the contextq
```
Comment (yesnault) [inline engine/api/workflow/dao_node.go]: context

File: engine/api/workflow/dao_hook_model.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+		return sdk.WrapError(sdk.ErrNotFound, "UpdateHookModel> Unable to update hook model %s", m.Name)
+	}
+	return nil
+}
+
+// InsertHookModel inserts a hook model in database
+func InsertHookModel(db gorp.SqlExecutor, m *sdk.WorkflowHookModel) error {
+	dbm := NodeHookModel(*m)
+	if err := db.Insert(&dbm); err != nil {
+		return sdk.WrapError(err, "UpdateHookModel> Unable to insert hook model %s", m.Name)
```
Comment (yesnault) [inline engine/api/workflow/dao_hook_model.go]: UpdateHookModel  -> InsertHookModel

File: engine/api/workflow/dao_hook_model.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+	return tx.Commit()
+}
+
+func checkBuiltinWorkflowHookModelExist(db gorp.SqlExecutor, h *sdk.WorkflowHookModel) (bool, error) {
+	var count = 0
+	if err := db.QueryRow("select count(1), id from workflow_hook_model where name = $1 group by id", h.Name).Scan(&count, &h.ID); err != nil {
+		if err == sql.ErrNoRows {
+			return false, nil
+		}
+		return false, err
```
Comment (sguiheux) [inline engine/api/workflow/dao_hook_model.go]: checkBuiltinWorkflowHookModelExist

File: engine/api/workflow/dao_hook_model.go [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+func LoadHookModels(db gorp.SqlExecutor) ([]sdk.WorkflowHookModel, error) {
+	dbModels := []NodeHookModel{}
+	if _, err := db.Select(&dbModels, "select id, name, type, image, command, author, description, identifier, icon from workflow_hook_model"); err != nil {
+		return nil, sdk.WrapError(err, "LoadHookModels> Unable to load WorkflowHookModel")
+	}
+	models := []sdk.WorkflowHookModel{}
+	for i := range dbModels {
+		m := dbModels[i]
+		if err := m.PostGet(db); err != nil {
+			return nil, err
```
Comment (sguiheux) [inline engine/api/workflow/dao_hook_model.go]: checkBuiltinWorkflowHookModelExist

File: engine/api/workflow/dao_hook_model.go [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+// LoadHookModelByName returns a hook model by it's name, if not founc, it returns an error
+func LoadHookModelByName(db gorp.SqlExecutor, name string) (*sdk.WorkflowHookModel, error) {
+	m := NodeHookModel{}
+	query := "select id, name, type, image, command, default_config, author, description, identifier, icon from workflow_hook_model where name = $1"
+	if err := db.SelectOne(&m, query, name); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, sdk.ErrNotFound
+		}
+		return nil, err
```
Comment (sguiheux) [inline engine/api/workflow/dao_hook_model.go]: checkBuiltinWorkflowHookModelExist

File: engine/api/workflow_hook.go [hunk 17]
```diff
@@ -0,0 +1,80 @@
+package main
+
+import (
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/ovh/cds/sdk"
```
Comment (sguiheux) [inline engine/api/workflow_hook.go]: import

File: engine/api/workflow/dao_hook_model.go [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+	return models, nil
+}
+
+// LoadHookModelByName returns a hook model by it's name, if not founc, it returns an error
+func LoadHookModelByName(db gorp.SqlExecutor, name string) (*sdk.WorkflowHookModel, error) {
+	m := NodeHookModel{}
+	query := "select id, name, type, image, command, default_config, author, description, identifier, icon from workflow_hook_model where name = $1"
+	if err := db.SelectOne(&m, query, name); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, sdk.WrapError(sdk.ErrNotFound, "LoadHookModels> Unable to load WorkflowHookModel")
```
Comment (yesnault) [inline engine/api/workflow/dao_hook_model.go]: LoadHookModels -> LoadHookModelByName

File: engine/api/workflow/dao_hook_model.go [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+		}
+		return nil, sdk.WrapError(err, "LoadHookModels> Unable to load WorkflowHookModel")
```
Comment (yesnault) [inline engine/api/workflow/dao_hook_model.go]: idem

**Pull request merged**